### PR TITLE
Add map_index to XCom's primary key and interface

### PR DESCRIPTION
**Pull request by uranusjr**
Since the latest XCom migration has not been publicly released, this
modifies that migration in place so we don't need to recreate the XCom
primary key *again*.

While I added the map_index argument to the functions, I'm honestly not
sure how it can be used for most of the interfaces. Those can hopefully
be added incrementally later.

### CODE REVIEW

**Comment by uranusjr** [comment]
Ah, adding `map_index` requires us to change the XCom interface (because the push/pull/clear interface now needs to accept `map_index` as an argument). This will need the patch-outdated treatment introduced to `serialize` in #19505.

File: airflow/models/xcom.py [hunk 1]
```diff
@@ -108,6 +111,7 @@ def set(
         dag_id: str,
         task_id: str,
         run_id: str,
+        map_index: int = -1,
```
Comment (ashb) [inline airflow/models/xcom.py]: I wonder if instead of all these columns we should take a `Union[TaskInstance, TaskInstanceKey]`?

**Review by ashb — APPROVED** [review]

File: airflow/models/xcom.py (hunk 1 above)
Comment (uranusjr) [reply airflow/models/xcom.py]: I was wondering about just `TaskInstanceKey`; it’s quite trivial to build a key from a ti anyway.

Comment (ashb) [reply airflow/models/xcom.py]: It's `ti.key` infact :)

Comment (ashb) [reply airflow/models/xcom.py]: Hmmm, I've started making this change and I'm not too sure about it. Will continue for now.

**Comment by ashb** [comment]
Decided to fix up the Operator links first, so this depends on https://github.com/apache/airflow/pull/21798

**Comment by uranusjr** [comment]
Easier to redo this one top of main instead of rebasing.

**Pull request closed without merging**